### PR TITLE
Add RFDiffusion cosine noise schedule

### PR DESCRIPTION
**Pull request by chidwipak**
## Description

Adds `CosineSchedule`, the diffusion noise schedule used by the RFDiffusion
protein backbone model.

The schedule implements the cosine variance schedule from Improved DDPM, which
tends to produce better samples than a linear schedule because the noise is
added more gradually at the start and end of the trajectory.

Three methods are exposed:

- `q_sample` – adds noise to clean data at a given timestep (the "forward" pass used during training).
- `p_sample` – denoises by one step using a model's noise prediction (the "reverse" pass used at inference).
- `sample` – runs the full reverse loop from noise to clean data.

All schedule tensors are stored as buffers so they move with the model when
you call `.to(device)` and are saved/loaded with checkpoints.

**Depends on PR #4994** (the embedding layers). This is PR B in a 4-part split.
PR C (BackboneDiffusion) and PR D (RFDiffusionModel) will build on top.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [x] Documentations (modification for documents)

## Checklist

- [x] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [ ] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
  - [ ] Run `mypy -p deepchem` and check no errors
  - [x] Run `flake8 <modified file> --count` and check no errors
  - [ ] Run `python -m doctest <modified file>` and check no errors
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: deepchem/models/torch_models/rfdiffusion.py [hunk 1]
```diff
@@ -185,3 +190,194 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+    >>> schedule = CosineSchedule(num_timesteps=100)
+    >>> import torch
+    >>> x0 = torch.randn(2, 10, 9)
+    >>> t = torch.tensor([0, 50])
+    >>> noisy_x, noise = schedule.q_sample(x0, t)
+    >>> noisy_x.shape
+    torch.Size([2, 10, 9])
+    """
+
+    def __init__(self, num_timesteps: int = 1000, s: float = 0.008) -> None:
```
Comment (JoseAntonioSiguenza) [inline deepchem/models/torch_models/rfdiffusion.py:241]: Is there an specific reason for setting default to 0.008? Does authors use this as default?

Comment (JoseAntonioSiguenza) [reply deepchem/models/torch_models/rfdiffusion.py:241]: I checked "Improved DDPM" paper, and `s` comes from pixel bin size of 1/127.5 ~ 0.00784 (by dividing the normalized range from -1 to 1 (range of 2) by 255 intervals). Please add this hyperparameter value to the documentation.

File: deepchem/models/torch_models/rfdiffusion.py [hunk 2]
```diff
@@ -185,3 +190,194 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+        super().__init__()
+        self.num_timesteps = num_timesteps
```
Comment (JoseAntonioSiguenza) [inline deepchem/models/torch_models/rfdiffusion.py:243]: Please add `num_timesteps` refers to "T" term from eq. 17 in reference paper

File: deepchem/models/torch_models/rfdiffusion.py [hunk 3]
```diff
@@ -185,3 +190,194 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+        alpha_cumprod = torch.cos((t + s) / (1 + s) * math.pi / 2)**2
+        alpha_cumprod = alpha_cumprod / alpha_cumprod[0]
+        betas = 1 - (alpha_cumprod[1:] / alpha_cumprod[:-1])
+        betas = torch.clamp(betas, 0, 0.999)
+
+        alphas = 1.0 - betas
+        alpha_cumprod = torch.cumprod(alphas, dim=0)
+        alpha_cumprod_prev = F.pad(alpha_cumprod[:-1], (1, 0), value=1.0)
+
+        self.register_buffer('betas', betas)
```
Comment (JoseAntonioSiguenza) [inline deepchem/models/torch_models/rfdiffusion.py]: I'm not sure about using `self.register_buffer` (I dont remember using it or see ir anywhere else in deepchem). I dont think this matches Deepchem's API. Is there an specific reason for registering buffer?

Comment (JoseAntonioSiguenza) [reply deepchem/models/torch_models/rfdiffusion.py]: I'd suggest creating a field to access from `q_sample`, `p_sample`, etc methods

File: deepchem/models/torch_models/rfdiffusion.py [hunk 4]
```diff
@@ -185,3 +190,194 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+        Parameters
+        ----------
+        x0 : torch.Tensor
+            Clean data of shape ``(batch, ...)``.
+        t : torch.Tensor
+            Integer timesteps of shape ``(batch,)``.
+        noise : torch.Tensor, optional
+            Pre-sampled noise tensor. Drawn from N(0, I) if not provided.
+
+        Returns
```
Comment (JoseAntonioSiguenza) [inline deepchem/models/torch_models/rfdiffusion.py:293]: Please add ussage examples

File: deepchem/models/torch_models/rfdiffusion.py [hunk 5]
```diff
@@ -185,3 +190,194 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+        sqrt_one_minus = self.sqrt_one_minus_alpha_cumprod[t_idx].to(x0.device)
+
+        while sqrt_alpha.dim() < x0.dim():
+            sqrt_alpha = sqrt_alpha.unsqueeze(-1)
+            sqrt_one_minus = sqrt_one_minus.unsqueeze(-1)
+
+        noisy_x = sqrt_alpha * x0 + sqrt_one_minus * noise
+        return noisy_x, noise
+
+    @torch.no_grad()
```
Comment (JoseAntonioSiguenza) [inline deepchem/models/torch_models/rfdiffusion.py]: I dont think `@torch.no_grad()` decorator matches Deepchem API. Instead please make it explicit inside method

File: deepchem/models/torch_models/rfdiffusion.py [hunk 6]
```diff
@@ -185,3 +190,194 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+
+        Parameters
+        ----------
+        model : nn.Module
+            The denoiser network. Called as ``model([x_t, t])``.
+        x_t : torch.Tensor
+            Noisy data of shape ``(batch, num_residues, coord_dim)``.
+        t : torch.Tensor
+            Current timestep of shape ``(batch,)``.
+
```
Comment (JoseAntonioSiguenza) [inline deepchem/models/torch_models/rfdiffusion.py:336]: Please add usage examples

File: deepchem/models/torch_models/rfdiffusion.py [hunk 7]
```diff
@@ -185,3 +190,194 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+        Returns
+        -------
+        torch.Tensor
+            Denoised data at timestep t-1, same shape as x_t.
+        """
+        device = x_t.device
+        noise_pred = model([x_t, t])
+
+        t_idx = t.to(self.betas.device)
```
Comment (JoseAntonioSiguenza) [inline deepchem/models/torch_models/rfdiffusion.py]: Can't we use `device` from ln 327? Since, `q_sample` also uses `.to(device)` lets add `self.device` in __init__ method.

File: deepchem/models/torch_models/rfdiffusion.py [hunk 8]
```diff
@@ -185,3 +190,194 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+        while nonzero_mask.dim() < x_t.dim():
+            nonzero_mask = nonzero_mask.unsqueeze(-1)
+
+        var = self.posterior_variance[t_idx].to(device)
+        while var.dim() < x_t.dim():
+            var = var.unsqueeze(-1)
+
+        return posterior_mean + nonzero_mask * torch.sqrt(var) * noise
+
+    @torch.no_grad()
```
Comment (JoseAntonioSiguenza) [inline deepchem/models/torch_models/rfdiffusion.py]: Same as above

File: deepchem/models/torch_models/rfdiffusion.py [hunk 9]
```diff
@@ -185,3 +190,194 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+        ----------
+        model : nn.Module
+            The denoiser network.
+        shape : tuple of int
+            Shape of samples to produce: ``(batch, num_residues, coord_dim)``.
+        device : torch.device
+            Device to generate on.
+
+        Returns
+        -------
```
Comment (JoseAntonioSiguenza) [inline deepchem/models/torch_models/rfdiffusion.py:401]: Please add usage examples

File: deepchem/models/torch_models/rfdiffusion.py [hunk 10]
```diff
@@ -185,3 +190,194 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+        var = self.posterior_variance[t_idx].to(device)
+        while var.dim() < x_t.dim():
+            var = var.unsqueeze(-1)
+
+        return posterior_mean + nonzero_mask * torch.sqrt(var) * noise
+
+    @torch.no_grad()
+    def sample(self, model: nn.Module, shape: Tuple[int, ...],
+               device: torch.device) -> torch.Tensor:
+        """Generate samples by running the full reverse diffusion loop.
```
Comment (JoseAntonioSiguenza) [inline deepchem/models/torch_models/rfdiffusion.py:389]: Please add how this will be used in RFDiffusion model

File: deepchem/models/torch_models/tests/test_rfdiffusion_schedule.py [hunk 11]
```diff
@@ -0,0 +1,86 @@
+"""Tests for the RFDiffusion cosine noise schedule."""
```
Comment (JoseAntonioSiguenza) [inline deepchem/models/torch_models/tests/test_rfdiffusion_schedule.py:1]: Please add brief docstrings to unit tests describing the purpose of each test (what errors ahould it catch, what are the expected outputs, etc)

File: deepchem/models/torch_models/tests/test_rfdiffusion_schedule.py [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+    has_torch = False
+
+
+@pytest.mark.torch
+@pytest.mark.skipif(not has_torch, reason="PyTorch not installed")
+class TestCosineSchedule:
+
+    def test_q_sample_shape(self):
+        schedule = CosineSchedule(num_timesteps=100)
+        x0 = torch.randn(2, 10, 9)
```
Comment (JoseAntonioSiguenza) [inline deepchem/models/torch_models/tests/test_rfdiffusion_schedule.py:21]: Lets add a test from real image data instead of random data. Can we test similar behaviour as original implementation? https://github.com/openai/improved-diffusion/tree/main/datasets

**Comment by JoseAntonioSiguenza** [comment]
Nice implementation! I left some comments related to Deepchem API and tests.

File: deepchem/models/torch_models/tests/test_rfdiffusion_schedule.py (hunk 12 above)
Comment (JoseAntonioSiguenza) [reply deepchem/models/torch_models/tests/test_rfdiffusion_schedule.py:21]: Lets test proper behaviour against real image data. I think if we succesfully this test, the implementation would be more robust and safe.

File: deepchem/models/torch_models/rfdiffusion.py (hunk 1 above)
Comment (chidwipak) [reply deepchem/models/torch_models/rfdiffusion.py:241]: > Is there an specific reason for setting default to 0.008? Does authors use this as default?

Added that context to the docstring. `0.008` is the default used in the Improved DDPM paper.

Comment (chidwipak) [reply deepchem/models/torch_models/rfdiffusion.py:241]: > I checked "Improved DDPM" paper, and `s` comes from pixel bin size of 1/127.5 ~ 0.00784 (by dividing the normalized range from -1 to 1 (range of 2) by 255 intervals). Please add this hyperparameter value to the documentation.

> Added that note next to `s`. The docstring now calls out the `1 / 127.5 ≈ 0.00784` bin-size motivation for the `0.008` default.

File: deepchem/models/torch_models/rfdiffusion.py (hunk 2 above)
Comment (chidwipak) [reply deepchem/models/torch_models/rfdiffusion.py:243]: Added that to the parameter docs. `num_timesteps` now points to the `T` term from equation 17.

File: deepchem/models/torch_models/rfdiffusion.py (hunk 3 above)
Comment (chidwipak) [reply deepchem/models/torch_models/rfdiffusion.py]: > I'm not sure about using `self.register_buffer` (I dont remember using it or see ir anywhere else in deepchem). I dont think this matches Deepchem's API. Is there an specific reason for registering buffer?

Kept `register_buffer` here because these tensors are fixed schedule state, not learnable weights. They still need to move with the module and be written to the checkpoint.

Comment (chidwipak) [reply deepchem/models/torch_models/rfdiffusion.py]: > I'd suggest creating a field to access from `q_sample`, `p_sample`, etc methods

These tensors are already exposed as fields through `register_buffer`, so the methods read them as `self.betas`, `self.alpha_cumprod`, and the rest.

File: deepchem/models/torch_models/rfdiffusion.py (hunk 4 above)
Comment (chidwipak) [reply deepchem/models/torch_models/rfdiffusion.py:293]: Added a short `q_sample()` example to the docstring.

File: deepchem/models/torch_models/rfdiffusion.py (hunk 5 above)
Comment (chidwipak) [reply deepchem/models/torch_models/rfdiffusion.py]: Changed both methods to use `with torch.no_grad():` inside the body so they follow the usual DeepChem style.

File: deepchem/models/torch_models/rfdiffusion.py (hunk 6 above)
Comment (chidwipak) [reply deepchem/models/torch_models/rfdiffusion.py:336]: Added a small `p_sample()` example with a dummy denoiser so the call pattern is clear.

File: deepchem/models/torch_models/rfdiffusion.py (hunk 7 above)
Comment (chidwipak) [reply deepchem/models/torch_models/rfdiffusion.py]: I kept the device on the tensors instead of adding a separate `self.device` field. These buffers already move with the module, and `q_sample()` only casts the indexed values to match the input tensor.

File: deepchem/models/torch_models/rfdiffusion.py (hunk 8 above)
Comment (chidwipak) [reply deepchem/models/torch_models/rfdiffusion.py]: I kept the same pattern in `sample()`. The schedule buffers already track the module device, so there is not a second device field to keep in sync.

File: deepchem/models/torch_models/rfdiffusion.py (hunk 9 above)
Comment (chidwipak) [reply deepchem/models/torch_models/rfdiffusion.py:401]: Added a short `sample()` example to the docstring.

File: deepchem/models/torch_models/rfdiffusion.py (hunk 10 above)
Comment (chidwipak) [reply deepchem/models/torch_models/rfdiffusion.py:389]: Added a short note on how `RFDiffusionModel` uses `q_sample()` during training and `sample()` during generation.

File: deepchem/models/torch_models/tests/test_rfdiffusion_schedule.py (hunk 11 above)
Comment (chidwipak) [reply deepchem/models/torch_models/tests/test_rfdiffusion_schedule.py:1]: Added brief docstrings to the schedule tests so each one says what behavior it is checking.

File: deepchem/models/torch_models/tests/test_rfdiffusion_schedule.py (hunk 12 above)
Comment (chidwipak) [reply deepchem/models/torch_models/tests/test_rfdiffusion_schedule.py:21]: I kept the test inputs synthetic and added a reference check against the Improved DDPM cosine formula instead. Since the schedule only depends on the timestep grid, that is closer to the original implementation than swapping in image tensors.

File: deepchem/models/torch_models/rfdiffusion.py (hunk 3 above)
Comment (JoseAntonioSiguenza) [reply deepchem/models/torch_models/rfdiffusion.py]: Yes, but my suggestion is to add the needed tensors(self.betas, self.alpha_cumprod, etc) explictily as fields instead of using `self.register_buffer`

**Comment by JoseAntonioSiguenza** [comment]
LGTM

**Review by rbharath — APPROVED** [review]
LGTM
